src/SpirvIR.cpp

```diff
@@ -0,0 +1,1710 @@
+}
+
+SpvId SpvBuilder::lookup_struct(const StructMemberTypes &member_type_ids) const {
+    TypeKey key = hash_struct(member_type_ids);
+    TypeMap::const_iterator it = struct_map.find(key);
+    if (it != struct_map.end()) {
+        return it->second;
+    }
+    return SpvInvalidId;
+}
+
+SpvId SpvBuilder::declare_struct(const StructMemberTypes &member_type_ids) {
+    TypeKey key = hash_struct(member_type_ids);
+    TypeMap::const_iterator it = struct_map.find(key);
+    if (it != struct_map.end()) {
+        return it->second;
+    }
+
+    SpvId struct_type_id = declare_id(SpvStructTypeId);
+    SpvInstruction inst = SpvFactory::struct_type(struct_type_id, member_type_ids);
+    module.add_type(inst);
+    struct_map[key] = struct_type_id;
+    return struct_type_id;
+}
+
+SpvBuilder::PointerTypeKey SpvBuilder::hash_pointer_type(const Type &type, SpvStorageClass storage_class) const {
+    SpvId base_type_id = lookup_type(type);
+    if (base_type_id == SpvInvalidId) {
+        internal_error << "SPIRV: Attempted to declare pointer type for undeclared base type! " << type << "\n";
+    }
+    return std::make_pair(base_type_id, storage_class);
+}
+
+SpvBuilder::PointerTypeKey SpvBuilder::hash_pointer_type(SpvId base_type_id, SpvStorageClass storage_class) const {
+    return std::make_pair(base_type_id, storage_class);
+}
+
+SpvId SpvBuilder::lookup_pointer_type(const Type &type, SpvStorageClass storage_class) const {
+    SpvId base_type_id = lookup_type(type);
+    if (base_type_id == SpvInvalidId) {
+        internal_error << "SPIRV: Attempted to lookup pointer type for undeclared base type! " << type << "\n";
+    }
+    return lookup_pointer_type(base_type_id, storage_class);
+}
+
+SpvId SpvBuilder::lookup_pointer_type(SpvId base_type_id, SpvStorageClass storage_class) const {
+    PointerTypeKey key = hash_pointer_type(base_type_id, storage_class);
+    PointerTypeMap::const_iterator it = pointer_type_map.find(key);
+    if (it != pointer_type_map.end()) {
+        return it->second;
+    }
+    return SpvInvalidId;
+}
+
+SpvId SpvBuilder::declare_pointer_type(const Type &type, SpvStorageClass storage_class) {
+    SpvId base_type_id = map_type(type);
+    return declare_pointer_type(base_type_id, storage_class);
+}
+
+SpvId SpvBuilder::declare_pointer_type(SpvId base_type_id, SpvStorageClass storage_class) {
+    PointerTypeKey key = hash_pointer_type(base_type_id, storage_class);
+    PointerTypeMap::const_iterator it = pointer_type_map.find(key);
+    if (it != pointer_type_map.end()) {
+        return it->second;
+    }
+
+    SpvId pointer_type_id = declare_id(SpvPointerTypeId);
+    SpvInstruction inst = SpvFactory::pointer_type(pointer_type_id, storage_class, base_type_id);
+    module.add_type(inst);
+    pointer_type_map[key] = pointer_type_id;
+    return pointer_type_id;
+}
+
+SpvBuilder::ConstantKey SpvBuilder::hash_constant(const Type &type, const void *data) const {
+    ConstantKey key(type.bytes() + 4, ' ');
```

Comment:
Yes, you're entirely correct ... the function name is misleading, since the hashing actually occurs in unordered_map. Unfortunately, none of the SPIR-V types have a fixed size (SPIR-V allows arbitrary vector widths and even arbitrary precision for integers and floats). We could adopt our own hashing scheme and produce a uint64_t value, but I assumed that std::hash would be more sophisticated.  A quick search reveals that libstdc++ uses MurmurHashUnaligned2 for its string hashing, which seems like a good choice.

Reply:
Perhaps renaming the fake hash methods into make_key methods would be better?  E.g.

`SpvBuilder::hash_constant() => SpvBuilder::make_constant_key()`